pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java

```diff
@@ -47,115 +67,422 @@
+      return;
+    }
+    Map<String, String> taskToCronExpressionMap = getTaskToCronExpressionMap(taskTypeConfigsMap);
+    LOGGER.info("Got taskToCronExpressionMap {} ", taskToCronExpressionMap);
+    updateCronTaskScheduler(tableWithType, taskToCronExpressionMap);
+  }
+
+  private void updateCronTaskScheduler(String tableWithType, Map<String, String> taskToCronExpressionMap) {
+    Map<String, String> existingScheduledTasks = _tableTaskTypeToCronExpressionMap.get(tableWithType);
+    if (existingScheduledTasks != null && !existingScheduledTasks.isEmpty()) {
+
+      // Loop over existing tasks to identify tasks to be removed or updated
+      for (Map.Entry<String, String> entry : existingScheduledTasks.entrySet()) {
+        String existingTaskType = entry.getKey();
+        String newCronExpression = taskToCronExpressionMap.get(existingTaskType);
+
+        if (newCronExpression == null) {
+          // Task should be removed
+          try {
+            _scheduler.deleteJob(JobKey.jobKey(tableWithType, existingTaskType));
+            _controllerMetrics.addValueToTableGauge(getCronJobName(tableWithType, existingTaskType),
+                ControllerGauge.CRON_SCHEDULER_JOB_SCHEDULED, -1L);
+          } catch (SchedulerException e) {
+            LOGGER.error("Failed to delete scheduled job for table {}, task type {}", tableWithType, existingTaskType,
+                e);
+          }
+          continue;
+        }
+
+        if (!entry.getValue().equalsIgnoreCase(newCronExpression)) {
+          // Update existing task with new cron expr
+          try {
+            TriggerKey triggerKey = TriggerKey.triggerKey(tableWithType, existingTaskType);
+            Trigger trigger = TriggerBuilder.newTrigger().withIdentity(triggerKey)
+                .withSchedule(CronScheduleBuilder.cronSchedule(newCronExpression)).build();
+            _scheduler.rescheduleJob(triggerKey, trigger);
+          } catch (SchedulerException e) {
+            LOGGER.error("Failed to update scheduled job for table {}, task type {}", tableWithType, existingTaskType,
+                e);
+          }
+        }
+      }
+
+      // Loop over new tasks to identify tasks to be added
+      for (Map.Entry<String, String> entry : taskToCronExpressionMap.entrySet()) {
+        String newTaskType = entry.getKey();
+        if (!existingScheduledTasks.containsKey(newTaskType)) {
+          String newCronExpression = entry.getValue();
+          try {
+            scheduleJob(tableWithType, newTaskType, newCronExpression);
+          } catch (SchedulerException e) {
+            LOGGER.error("Failed to schedule cron task for table {}, task {}, cron expr {}", tableWithType, newTaskType,
+                newCronExpression, e);
+          }
+        }
+      }
+    } else {
+      for (String taskType : taskToCronExpressionMap.keySet()) {
+        // Schedule new job
+        String cronExpr = taskToCronExpressionMap.get(taskType);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.